### PR TITLE
Fix cp38-macosx universal2 and arm64 Python pkg URL

### PR DESCRIPTION
**Pull request by hombit**
Previous URLs pointed to x86_64 version

### CODE REVIEW

**Comment by joerick** [comment]
Unfortunately we don't want to upgrade the bundled Python because it would increase the minimum MacOS version for the x86_64 arch from 10.9 to 11.0, which is a pretty massive jump.

File: cibuildwheel/resources/build-platforms.toml [hunk 1]
```diff
@@ -76,8 +76,8 @@ python_configurations = [
   { identifier = "cp36-macosx_x86_64", version = "3.6", url = "https://www.python.org/ftp/python/3.6.8/python-3.6.8-macosx10.9.pkg" },
   { identifier = "cp37-macosx_x86_64", version = "3.7", url = "https://www.python.org/ftp/python/3.7.9/python-3.7.9-macosx10.9.pkg" },
   { identifier = "cp38-macosx_x86_64", version = "3.8", url = "https://www.python.org/ftp/python/3.8.10/python-3.8.10-macosx10.9.pkg" },
-  { identifier = "cp38-macosx_arm64", version = "3.8", url = "https://www.python.org/ftp/python/3.8.10/python-3.8.10-macosx10.9.pkg" },
-  { identifier = "cp38-macosx_universal2", version = "3.8", url = "https://www.python.org/ftp/python/3.8.10/python-3.8.10-macosx10.9.pkg" },
+  { identifier = "cp38-macosx_arm64", version = "3.8", url = "https://www.python.org/ftp/python/3.8.10/python-3.8.10-macos11.pkg" },
+  { identifier = "cp38-macosx_universal2", version = "3.8", url = "https://www.python.org/ftp/python/3.8.10/python-3.8.10-macos11.pkg" },
```
[the remark this answers is not in the document]
Comment (hombit) [reply cibuildwheel/resources/build-platforms.toml:80]: Should I update arm64 URL only?

[the remark this answers is not in the document]
Comment (hombit) [reply cibuildwheel/resources/build-platforms.toml:80]: I still don't understand, sorry. For Python 3.9 and 3.10 both arm64 and universal2 URLs points to universal2 packages, while python 3.8 URLs are for the x86_64 package. Why is it different for different Python versions?

[the remark this answers is not in the document]
Comment (Czaki) [reply cibuildwheel/resources/build-platforms.toml:80]: I do not play too much with ARM on macOS so let's wait for what @joerick will tell.

**Comment by henryiii** [comment]
We tried very hard to use this when it came out, but it doesn’t work. CPython only has partial support for 3.8 on ARM - it was squeezed in at the last minute in the last binary release. IIRC you can’t produce 10.9-10.15 compatible wheels for Intel from the ARM version since Python didn’t include a 10.9 compatible Intel version in the binary, like they normally do (and do in 3.9+). You can look at the PR where 3.8 was added, I think.

I think to make this “work” we’d need to special case when running on ARM and have the (big) caveat that you can’t build for macOS 10 from AS (but only for 3.8). Is 3.8 still the version Apple ships? I’m traveling and won’t be my AS machine till the 27th.

**Comment by henryiii** [comment]
Read here: 

* https://github.com/pypa/cibuildwheel/pull/663#discussion_r628731708
* https://github.com/scipy/oldest-supported-numpy/pull/20#discussion_r641740734
* https://github.com/pypa/cibuildwheel/pull/704

I think we should see if we can build from AS too, but we haven’t had CI runners for it and it will take some changes and should error out of macOS version is set <11.

**Comment by mayeut** [comment]
> IIRC you can’t produce 10.9-10.15 compatible wheels for Intel from the ARM version since Python didn’t include a 10.9 compatible Intel version in the binary, like they normally do (and do in 3.9+).

Now that `MACOSX_DEPLOYMENT_TARGET` is always set, I don't think there's a big issue on what you can target (i.e. will still be 10.9 for x86_64) unless I'm missing something. However, using a macOS 10.x runner to do the build won't be possible anymore with this installer which requires macOS 11+.

**Comment by henryiii** [comment]
From what I understand, you can't set `MACOSX_DEPLOYMENT_TARGET` lower than whatever CPython was prepared with, otherwise we'd not need to download the official binaries at all. I could be wrong, but I'm pretty sure `MACOSX_DEPLOYMENT_TARGET` can only be increased.

**Comment by Czaki** [comment]
yes. here is a `wheel` code fragment with a warning about this:

https://github.com/pypa/wheel/blob/9d4264d2f23ea44dd85639a27fa6b284261921cd/src/wheel/macosx_libfile.py#L411-L419

**Comment by joerick** [comment]
Thanks all for the informative discussion. I guess the upshot is that we keep the macos10.9 installer, and we can't test 3.8 arm64 wheels, even when running on arm64. 

However, we should change the behaviour when running on Apple Silicon, to not try to test the arm64 3.8 wheel, and include an informative message explaining why.

**Comment by joerick** [comment]
Just a note for anyone finding this now that arm64 runners are more widely available. If it's important to you to test the cp38 arm64 wheels, there's a workaround.

You can choose to install a different version of Python 3.8 on your runner before invoking cibuildwheel:

```sh
curl -fsSLO https://www.python.org/ftp/python/3.8.10/python-3.8.10-macos11.pkg
sudo installer -pkg python-3.8.10-macos11.pkg -target /
# ... invoke cibuildwheel ...
```

Just be aware that if you use this Python to build x86_64 wheels too, this increases the minimum supported macOS version on x86_64 to 11.0.

**Comment by joerick** [comment]
For reference, this gives an idea of which Python 3.8 users are using which OS.

image

That's from our [stats spreadsheet](https://docs.google.com/spreadsheets/d/1jBYh6shr17DxzZZRoU0VjLA9s7UI95V-GJEFWl-q2L4/edit#gid=2008903908)

**Comment by henryiii** [comment]
Do you have your old python versions the last month plot? That was interesting to see the fraction of 3.8 ARM users. I also don't remember what the default Python version was for those versions. I'm on the latest macOS and I see 3.9 in my `/usr/bin`, so the've finally updated (a little).

**Comment by joerick** [comment]
image

Though I have no idea what's going on with the Python 2.7 arm64 users!? I guess that Homebrew offers a build of that?

**Comment by henryiii** [comment]
I don't think they do, I don't think there's an open source 2.7 for AS, and homebrew doesn't offer 2.7 anymore. But Apple provided their own patched version in macOS 11 - 12.3.

**Pull request closed without merging**